File: mobile/lib/models/answer_package.dart
// models/answer_package.dart

class AnswerPackage {
  final AnswerHeader header;
  final String payload; // Encrypted string
  final String signature;

  AnswerPackage({required this.header, required this.payload, required this.signature});

  Map<String, dynamic> toJson() => {
    'header': header.toJson(),
    'payload': payload,
    'signature': signature,
  };
}

class AnswerHeader {
  final String examId;
  final String studentId;
  final String attemptId;
  final String deviceId;
  final DateTime submitTime;
  final String iv;

  AnswerHeader({
    required this.examId,
    required this.studentId,
    required this.attemptId,
    required this.deviceId,
    required this.submitTime,
    required this.iv,
  });

  Map<String, dynamic> toJson() => {
    'exam_id': examId,
    'student_id': studentId,
    'attempt_id': attemptId,
    'device_id': deviceId,
    'submit_time': submitTime.toIso8601String(),
    'iv': iv,
  };
}

class AnswerPayload {
  final List<AnswerItem> answers;
  final List<AuditLog> logs;

  AnswerPayload({required this.answers, required this.logs});

  Map<String, dynamic> toJson() => {
    'answers': answers.map((e) => e.toJson()).toList(),
    'logs': logs.map((e) => e.toJson()).toList(),
  };
}

class AnswerItem {
  final String questionId;
  final dynamic answer; // string or list
  final int timeSpent;

  AnswerItem({required this.questionId, required this.answer, required this.timeSpent});

  Map<String, dynamic> toJson() => {
    'q_id': questionId,
    'ans': answer,
    'time_spent_sec': timeSpent,
  };
}

class AuditLog {
  final String event;
  final DateTime timestamp;

  AuditLog({required this.event, required this.timestamp});

  Map<String, dynamic> toJson() => {
    'event': event,
    'ts': timestamp.toIso8601String(),
  };
}

File: backend/models/answer_package.go
package models

import "time"

// AnswerPackage adalah struktur file .ans yang diupload siswa
type AnswerPackage struct {
	Header    AnswerHeader `json:"header"`
	Payload   string       `json:"payload"`   // Encrypted JSON of AnswerPayload
	Signature string       `json:"signature"` // Hash/Signature of payload
}

type AnswerHeader struct {
	ExamID      string    `json:"exam_id"`
	StudentID   string    `json:"student_id"`
	AttemptID   string    `json:"attempt_id"`
	DeviceID    string    `json:"device_id"`
	SubmitTime  time.Time `json:"submit_time"`
	IV          string    `json:"iv"`
}

type AnswerPayload struct {
	Answers []AnswerItem `json:"answers"`
	Logs    []AuditLog   `json:"logs"`
}

type AnswerItem struct {
	QuestionID string      `json:"q_id"`
	Answer     interface{} `json:"ans"` // string or []string
	TimeSpent  int         `json:"time_spent_sec"`
}

type AuditLog struct {
	Event     string    `json:"event"` // "app_focus_lost", "wifi_on"
	Timestamp time.Time `json:"ts"`
}
